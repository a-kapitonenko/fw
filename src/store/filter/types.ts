import { Frame } from '../frames/types';

export type Field = {
  name: string;
  value: string;
  checked?: boolean;
};

export type Groups = {
  color: Field[];
  width: Field[];
  noseBridge: Field[];
  shape: Field[];
  material: Field[];
};

export const enum FilterActionTypes {
  FETCH_REQUEST = '@@filter/FETCH_REQUEST',
  FETCH_SUCCESS = '@@filter/FETCH_SUCCESS',
  FETCH_ERROR = '@@filter/FETCH_ERROR',
  CHANGE_CHECKED = '@@filter/CHANGE_CHECKED',
};

export interface IFilterState {
  readonly fetching: boolean;
  readonly groups: Groups;
  readonly errors: string;
  readonly query: Object;
  readonly frames: Frame[];
};
