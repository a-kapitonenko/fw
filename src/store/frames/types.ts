export type Frame = {
  upc: number;
  label: string;
  value: string; 
  img: string;
  compatibility?: boolean;
};

export const enum FramesActionTypes {  
  SUBMIT_START = '@@frames/SUBMIT_START',
  SUBMIT_SUCCESS = '@@frames/SUBMIT_SUCCESS',
  SUBMIT_FAILED = '@@frames/SUBMIT_FAILED',
  FETCH_SIMILAR_FRAMES_START = '@@frames/FETCH_SIMILAR_FRAMES_START',
  FETCH_SIMILAR_FRAMES_SUCCESS = '@@frames/FETCH_SIMILAR_FRAMES_SUCCESS',
  FETCH_SIMILAR_FRAMES_FAILED = '@@frames/FETCH_SIMILAR_FRAMES_FAILED',
  CHECK_FRAME_START = '@@frames/CHECK_FRAME_START',
  CHECK_FRAME_SUCCESS = '@@frames/CHECK_FRAME_SUCCESS',
  CHECK_FRAME_FAILED = '@@frames/CHECK_FRAME_FAILED',
  OPEN = '@@frames/OPEN',
  CLOSE = '@@frames/CLOSE',
  SET_SELECTED_FRAME = '@@frames/SET_SELECTED_FRAME',
  CLEAR_SELECTED_FRAME = '@@frame/CLEAR_SELECTED_FRAME',
  CLEAR_RESULT = '@@frames/CLEAR_RESULT',
};

export interface FramesState {
  readonly isFetching: boolean;
  readonly errors: string;
  readonly open: boolean;
  readonly selectedFrame: Frame;
  readonly similarFrames: {
    readonly isFetching: boolean;
    readonly errors: string;
    readonly data: Frame[];
  };
};
