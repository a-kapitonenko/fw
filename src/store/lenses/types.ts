export type Lens = {
  name: string;
  value: number;
};

export const enum LensesActionTypes {
  FETCH_LENSES_START = '@@lenses/FETCH_LENSES_START',
  FETCH_LENSES_SUCCESS = '@@lenses/FETCH_LENSES_SUCCESS',
  FETCH_LENSES_FAILED = '@@lenses/FETCH_LENSES_FAILED',

  SAVE_LENS_START = '@@lenses/SAVE_LENS_START',
  SAVE_LENS_SUCCESS = '@@lenses/SAVE_LENS_SUCCESS',
  SAVE_LENS_FAILED = '@@lenses/SAVE_LENS_FAILED',

  CHECK_LENS_START = '@@lenses/CHECK_LENS_START',
  CHECK_LENS_SUCCESS = '@@lenses/CHECK_LENS_SUCCESS',
  CHECK_LENS_FALIED = '@@lenses/CHECK_LENS_FAILED',
};

export interface ILensesState {
  readonly isFetching: boolean;
  readonly errors: string;
  readonly lenses: Lens[];
};
