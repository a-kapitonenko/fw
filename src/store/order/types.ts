import { Frame } from '../frames/types';
import { Lens } from '../lenses/types';

export type OculusInfo = {
  sphere: string;
  cyclinder: string;
  axis: string;
  addition: string;
  PDDistance: string;
  PDNear: string;
  prism: string;
};

export type Prescription = {
  OD: OculusInfo;
  OS: OculusInfo;
};

export type Blueprint = {
  img: string;
};

export type Barcode = {
  img: string;
}

export const enum OrderActionTypes {
  SET_RX_INFORMATION = '@@order/SET_RX_INFORMATION',
  SET_RECOMMENDATION = '@@order/SET_RECOMMENDATION',
  SET_LENS = '@@order/SET_LENS',
  SET_FRAME = '@@order/SET_FRAME',
  SET_MESSAGE = '@@order/SET_MESSAGE',
  SET_FITTING_PROPERTIES = '@@order/SET_FITTING_PROPERTIES',
  SET_FITTING_HEIGHT = '@@order/SET_FITTING_HEIGHT',
  SET_BLUEPRINT = '@@order/SET_BLUEPRINT',
  SET_BARCODE = '@@order/SET_BARCODE',
  SET_ERRORS = '@@order/SET_ERRORS',
  DELETE_ERRORS = '@@order/DELETE_ERRORS'
};

export interface IOrderState {
  readonly prescription: Prescription;
  readonly frame: Frame;
  readonly fittingHeight: number;
  readonly fittingProperties: any;
  readonly lens: Lens;
  readonly recommendation: string;
  readonly message: string;
  readonly blueprint: Blueprint;
  readonly barcode: Barcode;
  readonly errors: object;
};
