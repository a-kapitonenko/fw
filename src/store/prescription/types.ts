export type OculusInfo = {
  sphere: string;
  cyclinder: string;
  axis: string;
  addition: string;
  PDDistance: string;
  PDNear: string;
  prism: string;
};

export type Rx = {
  OD: OculusInfo;
  OS: OculusInfo;
};

export interface IPrescriptionState {
  readonly isFetching: boolean;
  readonly errors: string;
  readonly rx: Rx;
  readonly fittingProperties: any;
  readonly fittingHeight: number;
}