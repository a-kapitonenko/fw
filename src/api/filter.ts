import request from './request';
import { getFilterGroups, getFilterFrames } from '../test/frames';

export const fetchGroupsRequest = () => {
  return request(getFilterGroups);
}

export const fetchFramesRequest = (data: any) => {
  return request(getFilterFrames, data);
};
