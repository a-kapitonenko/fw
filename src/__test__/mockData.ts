import { ApplicationState } from '../store';
import { OculusInfo, Prescription, Boss, Blueprint, Barcode, Errors } from '../store/order/types';
import { Frame, FramesState } from '../store/frames/types';
import { Groups, IFilterState } from '../store/filter/types';
import { Lens } from '../store/lenses/types';
import { initialValue } from '../test/database';
import { RouterState } from 'react-router-redux';
import { ISearchState } from '../store/search/types';

export const errorMessage: string = 'Error message';
export const recommendation: string = 'Based on the prescription, the lens selection will use an 8 base lenses that is 70 MM in diameter. We recommend you look for frames with a strong front curve and a small Frame ED (<60 ED)';
export const message = 'The selected parameters for frame and lens are acceptable for cotout.';
export const checkLensErrorMessage: string = "Selected lens doesn't compatibility with patient prescription. Please, select from the following lenses";
export const checkFrameErrorMessage: string = 'This lens does not accecptable with selected frame.';
export const valuesErrorMessage: string = 'Failed to load order values';

const oculusData: OculusInfo = {
  sphere: '1',
  cyclinder: '2',
  axis: '1',
  addition: '2',
  PDDistance: '1',
  PDNear: '1',
  prism: '1',
};

export const prescription: Prescription = {
  OD: oculusData,
  OS: oculusData,
};

export const lens: Lens = {
  name: 'Nikon Eyes* Advanced + Polycarbonate- Polarized Gray/Brown',
  value: 1,
};

export const lenses: Lens[] = [
  {
    name: 'Nikon Eyes* Advanced + Polycarbonate- Polarized Gray/Brown',
    value: 1
  },
  {
    name: 'Nikon Eyes* Advanced + Polycarbonate- Transitions* Signature VII',
    value: 2
  },
  {
    name: 'Nikon Eyes* Advanced + Polycarbonate- XTRActive* Gray',
    value: 3
  },
];

export const frame: Frame = {
  upc: 82850553738,
  label: 'FM17116',
  value: '82850553738',
  img: '1.png',
  compatibility: true,
};

export const frames: Frame[] = [
  {
    upc: 82850553738,
    label: 'FM17116',
    value: '82850553738',
    img: '1.png',
    compatibility: true,
  },
  {
    upc: 82850523999,
    label: 'SA17346',
    value: '82850523999',
    img: '2.png',
    compatibility: false,
  },
];

export const groups: Groups = {
  color: [
    { name: 'Black', value: 'black' },
  ],
  width: [
    { name: 'Narrow', value: 'narrow' },
  ],
  noseBridge: [
    { name: 'Standard', value: 'standard' },
  ],
  shape: [
    { name: 'Square', value: 'square' },
  ],
  material: [
    { name: 'Acetate', value: 'acetate' },
  ],
};

export const query: object = {
  color: ['black', 'gold', 'brown'],
  width: ['narrow', 'medium'],
  noseBridge: ['standard'],
  shape: ['square'],
  material: ['acetate'],
};

export const fittingHeight = [
  {
    name: '1',
    value: 1,
  },
  {
    name: '2',
    value: 2,
  },
];

export const blueprint: Blueprint = {
  img: '1.png',
};

export const barcode: Barcode = {
  img: '1.png',
};

export const boss: Boss = {} as Boss;
export const state: ApplicationState = {} as ApplicationState;
export const id: string = '6E8C896C14153342DE34BBF26F69A8135BA548864A8C838F035CB35A593C2399';
export const values: any = initialValue;
export const height: number = 3;

export const filledState: ApplicationState = {
  routing: {} as RouterState,
  frames: {} as FramesState,
  lenses: {
    isFetching: false,
    errors: '',
    lenses: lenses,
  },
  order: {
    isFetching: false,
    errors: <Errors>{},
    redirect: false,
    fittingProperties: [],
    recommendation: '',
    message: '',
    blueprint: <Blueprint>{},
    boss: <Boss>{
      prescription: prescription,
      fittingHeight: 0,
      frame: frame,
      lens: lens,
      barcode: <Barcode>{},
    },
  },
  filter: {} as IFilterState,
  search: {} as ISearchState,
};
