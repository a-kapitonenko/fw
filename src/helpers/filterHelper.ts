import { Groups, Field } from '../store/filter/types';

export function createFilterGroupsData(groups: Groups): Groups {
  const newGroups: Groups = <Groups>{};

  for (const group in groups) {
    newGroups[group] = groups[group].map((field: Field) => ({ ...field, checked: false }));
  }

  return newGroups;
}

export function isEmptyQuery(query: object): boolean {
  for (const item in query) {
    if (query[item].length !== 0) {
      return false;
    } 
  }

  return true;
}
