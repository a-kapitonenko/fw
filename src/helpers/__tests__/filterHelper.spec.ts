import { createFilterGroupsData, isEmptyQuery } from '../filterHelper';
import * as mockData from '../../mockData';
import { Groups } from '../../store/filter/types';

describe('filterHelper', () => {
  it('test createFilterGroupsData function', () => {
    const groups: Groups = mockData.groups;

    expect(createFilterGroupsData(groups)).toMatchSnapshot();
  });

  describe('test isEmptyQuery function', () => {
    const emptyQuery: object = {};
    const filledQuery: object = mockData.query;

    it('should return true for empty query', () => {
      expect(isEmptyQuery(emptyQuery)).toEqual(true);
    });

    it('should return false for filled query', () => {
      expect(isEmptyQuery(filledQuery)).toEqual(false);
    });
  });
});
