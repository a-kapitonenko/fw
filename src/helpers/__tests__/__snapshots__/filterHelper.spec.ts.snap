// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterHelper test createFilterGroupsData function 1`] = `
Object {
  "color": Array [
    Object {
      "checked": false,
      "name": "Black",
      "value": "black",
    },
  ],
  "material": Array [
    Object {
      "checked": false,
      "name": "Acetate",
      "value": "acetate",
    },
  ],
  "noseBridge": Array [
    Object {
      "checked": false,
      "name": "Standard",
      "value": "standard",
    },
  ],
  "shape": Array [
    Object {
      "checked": false,
      "name": "Square",
      "value": "square",
    },
  ],
  "width": Array [
    Object {
      "checked": false,
      "name": "Narrow",
      "value": "narrow",
    },
  ],
}
`;
